### PR TITLE
SSR: Cache State, Take Two

### LINKED ISSUE #13352: Themes: DocumentHead tags are lost on second hit
In logged-out mode, we compute `<title />`, `<meta />`, and `<link />` tags (mostly for SEO) [here](Themes: DocumentHead tags are lost on second hit). However, as @seear found out, while the markup that we use to render our React component tree is [cached](https://github.com/Automattic/wp-calypso/blob/6b81c2d29563b106aeac37acdf1c484427995341/server/render/index.js#L49-L52), those fields aren't. This means that only the first time any SSR'd route is hit, the rendered page contains those HTML tags. On all subsequent hits, they aren't included!

To reproduce:
* Start a local Calypso instance (hard to reproduce on wordpress.com due to interfering hits by users)
* In an incognito window, land on http://calypso.localhost:3000/themes
* Check Page Source
* Verify that it contains a `<meta name="description" />` tag
* Reload, verify that tag is gone.
* Try the same e.g. with http://calypso.localhost:3000/theme/twentyseventeen

Instead of monkey-patching markup cache to also include non-markup stuff, I'm opting for a cleaner solution: #13339

### PR DESCRIPTION
**Pull request by ockham**
Supersedes #10988, and fixes #13352.

_tl;dr: Simplify server-side API data caching by leveraging Redux state serialization._

Previously, controllers of SSR'd sections were responsible for caching API data. But there's a better way: We're already sending Redux state to the client at server-render time, so why not also cache on the server side, and use that as the initial state when creating the Redux store for the next request (given that this is in logged-out mode, i.e. the cache key -- the route -- maps one-to-one to the state, with the exception of errors like 404s, where one error page can be enough for different invalid routes).

Furthermore, this PR removes `renderCacheKey` and just uses `context.pathname` for _markup_ cache. Previously, we were using concatenated path and API data (~state) cache timestamp as key. Creating the markup cache key isn't the section controller's responsibility now. (If we still need the timestamp, we can add it back when caching in `server/render`.)

To test:
* Restart Calypso using `DEBUG="calypso:server-render" make run`
* Verify that the theme showcase still works as before. 
* Verify that the 404 page for inexistent themes only takes up one cache slot: 
  * Logged-out, land on http://calypso.localhost:3000/theme/blah. You should get a 404
  * In the node console, you should see that your request triggered a `calypso:server-render cache miss for key +0ms Theme Not Found` (among other things)
  * Land on http://calypso.localhost:3000/theme/bleh -- no cache miss!
* Most importantly, verify that #13352 is fixed!

Questions:
* Do we want to increase [cache size](https://github.com/Automattic/wp-calypso/pull/13339/files#diff-43858cb8bd16f173b243a0584f1f83b2R8) now?
* Do we need to add back time stamps to markup cache keys? I'm not sure the old timestamps were really doing anything, since we never expired them AFAICS, so it seems we were effectively just relying on the LRU.

### CODE REVIEW

File: server/pages/index.js [hunk 1]
```diff
@@ -38,6 +40,12 @@ const staticFiles = [
 
 const sections = sectionsModule.get();
 
+function getInitialServerState( serializedServerState ) {
+	// Bootstrapped state from a server-render
+	const serverState = reducer( serializedServerState, { type: DESERIALIZE } );
+	return pick( serverState, Object.keys( serializedServerState ) );
```
Comment (ockham) [inline server/pages/index.js]: Carrying over @mcsf's https://github.com/Automattic/wp-calypso/pull/10988#discussion_r99832419 (which I never answered there 😳 but will here):

> Looking at this function and `getDefaultContext`, it's not clear to me why it seems that we:
> 1. send `serializeCachedServerState` through a `DESERIALIZE` reduction
> 2. pick the values to match the shape of that initial cached state
> 3. send the picked state again through a `DESERIALIZE` reduction

> My questions would be:
> - What is `pick` effectively weeding out?
> - Why the apparent double deserialization?

Comment (ockham) [inline server/pages/index.js]: Good spot, should be fixed by 05db4cbc0665fc74343b0822f4d3fe1c12f50afc.

Comment (ockham) [inline server/pages/index.js]: > What is `pick` effectively weeding out?

Currently, we only persist stuff on the server that we need for SSR (`document-head`, `themes`, and `ui` subtrees). We omit anything else because we don't need it, and it might unnecessarily bloat our cache.

File: server/render/index.js [hunk 2]
```diff
@@ -17,6 +17,9 @@ import {
 	getDocumentHeadMeta,
 	getDocumentHeadLink
 } from 'state/document-head/selectors';
+import { reducer } from 'state';
+import { SERIALIZE } from 'state/action-types';
+import stateCache from 'state-cache';
 
 const debug = debugFactory( 'calypso:server-render' );
 const markupCache = new Lru( { max: 3000 } );
```
Comment (seear) [inline server/render/index.js]: We should give this cache a max age now that the keys are not timestamped (previously the keys only lasted for an hour, meaning entries would drop out of the LRU after that time).

**Comment by seear** [comment]
There is problem with using `context.pathname` as the cache key. On `master` a request for `/themes?s=blue` skips the middlewares because of the query arg, and falls back to the stringified layout for the markup cache, where it caches an empty page with a masterbar:

```
  calypso:server-render cache access for key +8s {"key":null,"ref":null,"props":{"store":{}},"_owner":null,"_store":{}}
  calypso:server-render Server render time (ms) +0ms 0
GET /themes?s=blue 304 307.975 ms - -
```

On this branch, a request for `/themes?s=blue` caches the empty page with the key `/themes`:
```
  calypso:server-render cache access for key +0ms /themes
  calypso:server-render cache miss for key +1ms /themes
  calypso:server-render Server render time (ms) +33ms 34
GET /themes?s=blue 200 3626.222 ms - 63652
```
Meaning that any subsequent request to /themes (without the query arg) serves up an empty page.

If we fixed this by using the full URL including the query, it would use a cache slot for every different query that produces an empty page.

**Comment by ockham** [comment]
Good catch. What I wanted to achieve was to not cache routes with a querystring at all. I guess I'm missing a check for that somewhere.

**Comment by seear** [comment]
> where it caches an empty page with a masterbar

SSRing an empty page with a master bar for logged-out routes was added as a half-way step to full SSRd pages. It might be a good simplification to only SSR complete pages.

**Comment by ockham** [comment]
I just pushed a commit to disable caching for query strings -- in my testing, this seems to have fixed the incorrect lookups.

File: server/render/index.js [hunk 3]
```diff
@@ -91,7 +100,14 @@ export function serverRender( req, res ) {
 			reduxSubtrees = reduxSubtrees.concat( [ 'ui', 'themes' ] );
 		}
 
+		// Send state to client. Don't we need to serialize here?
```
Comment (seear) [inline server/render/index.js]: stringified [here](https://github.com/Automattic/wp-calypso/blob/c8e7f1d989129bdbeb3007e28a1e74ffc937f81e/server/pages/index.jade#L126)

File: server/pages/index.js [hunk 4]
```diff
@@ -111,6 +120,14 @@ function getCurrentCommitShortChecksum() {
 }
 
 function getDefaultContext( request ) {
+	let initialServerState = {};
+	// We don't cache routes with query params
+	if ( isEmpty( request.query ) ) {
```
Comment (seear) [inline server/pages/index.js]: Seems to me that this decision should be left up to the route, but I think we can go with this for now.

File: server/render/index.js [hunk 5]
```diff
@@ -76,9 +81,13 @@ export function serverRender( req, res ) {
...
 	}
 
-	if ( config.isEnabled( 'server-side-rendering' ) && context.layout && ! context.user ) {
-		const key = context.renderCacheKey || JSON.stringify( context.layout );
-		Object.assign( context, render( context.layout, key ) );
+	if ( config.isEnabled( 'server-side-rendering' ) && context.layout && ! context.user && isEmpty( context.query ) ) {
+		// context.pathname doesn't include querystring, so it's a suitable cache key.
+		let key = context.pathname;
+		if ( req.error ) {
+			key = req.error.message;
```
Comment (seear) [inline server/render/index.js]: neato

**Comment by seear** [comment]
/login route added in #13035 works fine

**Comment by seear** [comment]
> Do we want to increase cache size now?

I don't think so—we don't expect this PR cause more slots to be used, do we? We should get some logging soon via D5256-code.

**Comment by ockham** [comment]
Thanks! :slightly_smiling_face:
